nengo/ensemble.py

```diff
@@ -108,8 +108,8 @@ class Ensemble(NengoObject):
 
     probeable = ('decoded_output', 'input', 'scaled_encoders')
 
-    n_neurons = IntParam('n_neurons', default=None, low=1)
-    dimensions = IntParam('dimensions', default=None, low=1)
+    n_neurons = IntParam('n_neurons', default=1, low=1)
+    dimensions = IntParam('dimensions', default=1, low=1)
```

Comment:
These parameters aren't optional, are they? Does setting the default have any effect for non-optional parameters?

Reply:
You could in theory create an ensemble like `nengo.Ensemble(Default, Default)`, which would use the default values specified here (previously this would have tried to use `None`, resulting in an error).  We could also make these parameters `Unconfigurable` if we wanted.

Reply:
`nengo.Ensemble(Default, Default)` doesn't seem like something I would want to do or encourage. Not sure at the moment what the implications of `Unconfigurable` are (need to read the related code first).

I think `n_neurons` should not have a default value because it is not clear what a sensible default would be. 1 is the lower bound, but it is rare to actually want an ensemble with just one neuron. To me it seems better to force the user to be explicit.

I am more ok with the default `dimensions=1`, but would then go all the way and make the parameter optional to allow `nengo.Ensemble(n_neurons)`.

In either case I would suggest bringing those changes up in the dev meeting if we want to go along with them. They seem somewhat fundamental. (Or just leave as it is now with `default=None`.)

Reply:
Just to be clear, `nengo.Ensemble(Default, Default)` isn't a new feature introduced by this PR, that's always been possible.  So there isn't any fundamentally new feature being introduced here; it was just bugged before, because the default value wasn't a valid value.  So we probably shouldn't leave it as `default=None`.  

Setting it to `Unconfigurable` would prevent `nengo.Ensemble(Default, Default)`, which would be a minor change in behaviour but pretty unlikely to affect anyone.  I'd be slightly inclined to leave it as a default of 1, just because it allows more functionality and there doesn't seem to be any downsides (a user can't accidentally end up confused about the default value, since they have to explicitly request it).  But I don't think it matters too much either way, since this is a pretty edge use case.  We can definitely talk about it more in the dev meeting if you want to add it to the agenda though :+1:.

Reply:
Here is my thought process: First I'm asking myself should these parameters have a default value? This comes basically down to whether there is some specific value with a special property (e.g. a transform of 1 is implementing an identity transform) or we want to say some value is a good value to use (e.g. firing rates between 200 and 400 Hz). Let us assume there is such a value for `n_neurons` (or `dimensions`), then there should not be a problem with going all the way and make the constructor argument itself optional.

But I think there is no reasonable default value for `n_neurons`. Yes, `n_neurons=1` is special in so far as it is the lower bound, but advertising this as a good default value seems wrong to me. Yes, it would allow one to do `nengo.Ensemble(Default, Default)`, but I don't see why someone would want to do that instead of `nengo.Ensemble(1, 1)` (if someone actually wants a single neuron). What I can see a use case for is something like

```python
with nengo.Config(nengo.Ensemble) as cfg:
    cfg[nengo.Ensemble].n_neurons = 50
    cfg[nengo.Ensemble].dimensions = 1
    a = nengo.Ensemble(Default, Default)
    b = nengo.Ensemble(Default, Default)
    # ...
```

That, however, already works even with a default of `n_neurons=None` and I'm ok with that because the user still explicitely sets the number of neurons. It is a bit weird that someone has to pass in `Default` explicitely (I am not aware of any other situations in the API where that would be the case because usually the constructor arguments get a default of `Default`) and I might argue that the better way to set `n_neurons` on a bunch of ensemble would be:

```python
n_neurons = 50
dimenisons = 1
a = nengo.Ensemble(n_neurons, dimensions)
b = nengo.Ensemble(n_neurons, dimensions)
```

If this is in a subnetwork and someone wants to configure the neuron number from outside, it either needs to be passed in or it needs to be set after the creation of the network. That might seem less powerful than the Config approach, but that approach requires the subnetwork to already pass `Default` as argument to `n_neurons` which than requires the network to be created in the context of an appropriate config. It seems that explicitely requesting the number of neurons for that network as a constructor argument (to make that dependency clear) is a better idea. Hence, it seems to me that there is not much of a good use case for `nengo.Ensemble(Default, Default)` and we might as well set `n_neurons=Unconfigurable`.

Technically, setting `n_neurons=Unconfigurable` is a breaking change and I don't think there is anything in the current behaviour to be bothered by (it just enforces that the user sets `n_neurons` explicitely). So I would be fine either keeping current behaviour or making `n_neurons` `Unconfigurable` (as it is unlikely to actually affect anyone).

One final point about about setting `n_neurons=1` (or to some other default), but not making it optional: To me it seems that this is hard to explain/document which I usually take as an indication that there might be some better way to structure things. Because we were to set a default value, we ought to document it in the docstring, but at the same time function arguments with a default are also optional in Python. So I am not sure how to document this behaviour that there is a default value, but it only comes into play in this special circumstance where the function argument is set explicitely to the `Default`.

**tl;dr:**

* I am not convinced that there is a good use case for `nengo.Ensemble(Default, Default)`.
* I do not think that there is a good default value for `n_neurons`.
* If we set a default anyways, we should also make the parameter optional.
* I'm fine with either keeping current behaviour or change the default to `Unconfigurable` (technically breaking, but unlikely to affect anyone).